File: interfaz1.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

class Autor {
    private String nombre;
    private String apellido;

    public Autor(String nombre, String apellido) {
        this.nombre = nombre;
        this.apellido = apellido;
    }

    public String mostrarInfo() {
        return "Autor: " + nombre + " " + apellido;
    }
}

class Categoria {
    private String nombre;

    public Categoria(String nombre) {
        this.nombre = nombre;
    }

    public String mostrarInfo() {
        return "Categoría: " + nombre;
    }
}

class Libro {
    private String titulo;
    private String isbn;
    private Autor autor;
    private Categoria categoria;

    public Libro(String titulo, String isbn, Autor autor, Categoria categoria) {
        this.titulo = titulo;
        this.isbn = isbn;
        this.autor = autor;
        this.categoria = categoria;
    }

    public String mostrarInfo() {
        return "Libro: " + titulo + ", ISBN: " + isbn + "\n" + autor.mostrarInfo() + "\n" + categoria.mostrarInfo();
    }

    public String getTitulo() {
        return titulo;
    }
}

class Usuario {
    private String nombre;
    private String apellido;
    private String idUsuario;

    public Usuario(String nombre, String apellido, String idUsuario) {
        this.nombre = nombre;
        this.apellido = apellido;
        this.idUsuario = idUsuario;
    }

    public String mostrarInfo() {
        return "Usuario: " + nombre + " " + apellido + ", ID: " + idUsuario;
    }

    public String getIdUsuario() {
        return idUsuario;
    }
}

class Prestamo {
    private Libro libro;
    private Usuario usuario;
    private Date fechaPrestamo;
    private Date fechaDevolucion;

    public Prestamo(Libro libro, Usuario usuario, Date fechaPrestamo) {
        this.libro = libro;
        this.usuario = usuario;
        this.fechaPrestamo = fechaPrestamo;
    }

    public String mostrarInfo() {
        String info = "Préstamo del libro '" + libro.getTitulo() + "' a " + usuario.mostrarInfo() + "\n";
        info += "Fecha de préstamo: " + fechaPrestamo + "\n";
        if (fechaDevolucion != null) {
            info += "Fecha de devolución: " + fechaDevolucion + "\n";
        } else {
            info += "El libro no ha sido devuelto.\n";
        }
        return info;
    }

    public Libro getLibro() {
        return libro;
    }

    public Usuario getUsuario() {
        return usuario;
    }

    public void devolver() {
        fechaDevolucion = new Date();
    }
}

class Biblioteca {
    private List<Libro> libros;
    private List<Usuario> usuarios;
    private List<Prestamo> prestamos;

    public Biblioteca() {
        libros = new ArrayList<>();
        usuarios = new ArrayList<>();
        prestamos = new ArrayList<>();
    }

    public void registrarLibro(Libro libro) {
        libros.add(libro);
        JOptionPane.showMessageDialog(null, "Libro '" + libro.getTitulo() + "' registrado en la biblioteca.");
    }

    public void registrarUsuario(Usuario usuario) {
        usuarios.add(usuario);
        JOptionPane.showMessageDialog(null, "Usuario '" + usuario.mostrarInfo() + "' registrado en la biblioteca.");
    }

    public void realizarPrestamo(Libro libro, Usuario usuario) {
        Prestamo prestamo = new Prestamo(libro, usuario, new Date());
        prestamos.add(prestamo);
        JOptionPane.showMessageDialog(null, "Préstamo realizado: '" + libro.getTitulo() + "' a " + usuario.mostrarInfo());
    }

    public void devolverLibro(Prestamo prestamo) {
        prestamo.devolver();
        JOptionPane.showMessageDialog(null, "Libro '" + prestamo.getLibro().getTitulo() + "' devuelto por " + prestamo.getUsuario().mostrarInfo());
    }

    public void mostrarLibros() {
        StringBuilder info = new StringBuilder();
        if (!libros.isEmpty()) {
            info.append("Libros en la biblioteca:\n");
            for (Libro libro : libros) {
                info.append(libro.mostrarInfo()).append("\n\n");
            }
            JOptionPane.showMessageDialog(null, info.toString());
        } else {
            JOptionPane.showMessageDialog(null, "No hay libros registrados en la biblioteca.");
        }
    }

    public Libro buscarLibro(String titulo) {
        for (Libro libro : libros) {
            if (libro.getTitulo().equalsIgnoreCase(titulo)) {
                return libro;
            }
        }
        return null;
    }

    public Usuario buscarUsuario(String idUsuario) {
        for (Usuario usuario : usuarios) {
            if (usuario.getIdUsuario().equalsIgnoreCase(idUsuario)) {
                return usuario;
            }
        }
        return null;
    }

    public Prestamo buscarPrestamo(String titulo, String idUsuario) {
        for (Prestamo prestamo : prestamos) {
            if (prestamo.getLibro().getTitulo().equalsIgnoreCase(titulo) && prestamo.getUsuario().getIdUsuario().equalsIgnoreCase(idUsuario)) {
                return prestamo;
            }
        }
        return null;
    }
}

public class interfaz1 extends JFrame {
    private Biblioteca biblioteca;

    private JTextField tituloLibroField, isbnField, autorField, categoriaField;
    private JTextField nombreUsuarioField, apellidoUsuarioField, idUsuarioField;
    private JTextField tituloPrestamoField, idUsuarioPrestamoField;
    private JTextField tituloDevolucionField, idUsuarioDevolucionField;

    public interfaz1() {
        biblioteca = new Biblioteca();

        setTitle("Biblioteca");
        setSize(400, 600);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(new GridLayout(20, 2));

        // Widgets para registrar libros
        add(new JLabel("Registrar Libro"));
        add(new JLabel(""));
        add(new JLabel("Título"));
        tituloLibroField = new JTextField();
        add(tituloLibroField);
        add(new JLabel("ISBN"));
        isbnField = new JTextField();
        add(isbnField);
        add(new JLabel("Autor"));
        autorField = new JTextField();
        add(autorField);
        add(new JLabel("Categoría"));
        categoriaField = new JTextField();
        add(categoriaField);
        JButton registrarLibroBtn = new JButton("Registrar Libro");
        registrarLibroBtn.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                registrarLibro();
            }
        });
        add(registrarLibroBtn);
        add(new JLabel(""));

        // Widgets para registrar usuarios
        add(new JLabel("Registrar Usuario"));
        add(new JLabel(""));
        add(new JLabel("Nombre"));
        nombreUsuarioField = new JTextField();
        add(nombreUsuarioField);
        add(new JLabel("Apellido"));
        apellidoUsuarioField = new JTextField();
        add(apellidoUsuarioField);
        add(new JLabel("ID Usuario"));
        idUsuarioField = new JTextField();
        add(idUsuarioField);
        JButton registrarUsuarioBtn = new JButton("Registrar Usuario");
        registrarUsuarioBtn.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                registrarUsuario();
            }
        });
        add(registrarUsuarioBtn);
        add(new JLabel(""));

        // Widgets para realizar préstamos
        add(new JLabel("Realizar Préstamo"));
        add(new JLabel(""));
        add(new JLabel("Título Libro"));
        tituloPrestamoField = new JTextField();
        add(tituloPrestamoField);
        add(new JLabel("ID Usuario"));
        idUsuarioPrestamoField = new JTextField();
        add(idUsuarioPrestamoField);
        JButton realizarPrestamoBtn = new JButton("Realizar Préstamo");
        realizarPrestamoBtn.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                realizarPrestamo();
            }
        });
        add(realizarPrestamoBtn);
        add(new JLabel(""));

        // Widgets para devolver libros
        add(new JLabel("Devolver Libro"));
        add(new JLabel(""));
        add(new JLabel("Título Libro"));
        tituloDevolucionField = new JTextField();
        add(tituloDevolucionField);
        add(new JLabel("ID Usuario"));
        idUsuarioDevolucionField = new JTextField();
        add(idUsuarioDevolucionField);
        JButton devolverLibroBtn = new JButton("Devolver Libro");
        devolverLibroBtn.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                devolverLibro();
            }
        });
        add(devolverLibroBtn);
        add(new JLabel(""));

        // Widget para mostrar libros
        JButton mostrarLibrosBtn = new JButton("Mostrar Libros");
        mostrarLibrosBtn.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                mostrarLibros();
            }
        });
        add(mostrarLibrosBtn);
        add(new JLabel(""));
    }

    private void registrarLibro() {
        String titulo = tituloLibroField.getText();
        String isbn = isbnField.getText();
        String[] autorNombreApellido = autorField.getText().split(" ");
        String categoriaNombre = categoriaField.getText();

        Autor autor = new Autor(autorNombreApellido[0], autorNombreApellido[1]);
        Categoria categoria = new Categoria(categoriaNombre);
        Libro libro = new Libro(titulo, isbn, autor, categoria);

        biblioteca.registrarLibro(libro);
    }

    private void registrarUsuario() {
        String nombre = nombreUsuarioField.getText();
        String apellido = apellidoUsuarioField.getText();
        String idUsuario = idUsuarioField.getText();

        Usuario usuario = new Usuario(nombre, apellido, idUsuario);

        biblioteca.registrarUsuario(usuario);
    }

    private void realizarPrestamo() {
        String titulo = tituloPrestamoField.getText();
        String idUsuario = idUsuarioPrestamoField.getText();

        Libro libro = biblioteca.buscarLibro(titulo);
        Usuario usuario = biblioteca.buscarUsuario(idUsuario);

        if (libro != null && usuario != null) {
            biblioteca.realizarPrestamo(libro, usuario);
        } else {
            JOptionPane.showMessageDialog(this, "Libro o Usuario no encontrado.");
        }
    }

    private void devolverLibro() {
        String titulo = tituloDevolucionField.getText();
        String idUsuario = idUsuarioDevolucionField.getText();

        Prestamo prestamo = biblioteca.buscarPrestamo(titulo, idUsuario);

        if (prestamo != null) {
            biblioteca.devolverLibro(prestamo);
        } else {
            JOptionPane.showMessageDialog(this, "Préstamo no encontrado.");
        }
    }

    private void mostrarLibros() {
        biblioteca.mostrarLibros();
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                new interfaz1().setVisible(true);
            }
        });
    }
}

File: interfaz1.py
import tkinter as tk
from tkinter import messagebox
from datetime import date

class Autor:
    def __init__(self, nombre, apellido):
        self.nombre = nombre
        self.apellido = apellido

    def mostrar_info(self):
        return f"Autor: {self.nombre} {self.apellido}"

class Categoria:
    def __init__(self, nombre):
        self.nombre = nombre

    def mostrar_info(self):
        return f"Categoría: {self.nombre}"

class Libro:
    def __init__(self, titulo, isbn, autor, categoria):
        self.titulo = titulo
        self.isbn = isbn
        self.autor = autor
        self.categoria = categoria

    def mostrar_info(self):
        return f"Libro: {self.titulo}, ISBN: {self.isbn}\n{self.autor.mostrar_info()}\n{self.categoria.mostrar_info()}"

class Usuario:
    def __init__(self, nombre, apellido, id_usuario):
        self.nombre = nombre
        self.apellido = apellido
        self.id_usuario = id_usuario

    def mostrar_info(self):
        return f"Usuario: {self.nombre} {self.apellido}, ID: {self.id_usuario}"

class Prestamo:
    def __init__(self, libro, usuario, fecha_prestamo, fecha_devolucion=None):
        self.libro = libro
        self.usuario = usuario
        self.fecha_prestamo = fecha_prestamo
        self.fecha_devolucion = fecha_devolucion

    def mostrar_info(self):
        info = f"Préstamo del libro '{self.libro.titulo}' a {self.usuario.nombre} {self.usuario.apellido}\n"
        info += f"Fecha de préstamo: {self.fecha_prestamo}\n"
        if self.fecha_devolucion:
            info += f"Fecha de devolución: {self.fecha_devolucion}\n"
        else:
            info += "El libro no ha sido devuelto.\n"
        return info

class Biblioteca:
    def __init__(self):
        self.libros = []
        self.usuarios = []
        self.prestamos = []

    def registrar_libro(self, libro):
        self.libros.append(libro)
        messagebox.showinfo("Registro de Libro", f"Libro '{libro.titulo}' registrado en la biblioteca.")

    def registrar_usuario(self, usuario):
        self.usuarios.append(usuario)
        messagebox.showinfo("Registro de Usuario", f"Usuario '{usuario.nombre} {usuario.apellido}' registrado en la biblioteca.")

    def realizar_prestamo(self, libro, usuario):
        prestamo = Prestamo(libro, usuario, date.today())
        self.prestamos.append(prestamo)
        messagebox.showinfo("Préstamo Realizado", f"Préstamo realizado: '{libro.titulo}' a {usuario.nombre} {usuario.apellido}")

    def devolver_libro(self, prestamo):
        prestamo.fecha_devolucion = date.today()
        messagebox.showinfo("Devolución de Libro", f"Libro '{prestamo.libro.titulo}' devuelto por {prestamo.usuario.nombre} {prestamo.usuario.apellido}")

    def mostrar_libros(self):
        if self.libros:
            info = "Libros en la biblioteca:\n"
            for libro in self.libros:
                info += libro.mostrar_info() + "\n\n"
            messagebox.showinfo("Libros en la Biblioteca", info)
        else:
            messagebox.showinfo("Libros en la Biblioteca", "No hay libros registrados en la biblioteca.")

# Crear la interfaz gráfica
class BibliotecaApp:
    def __init__(self, root):
        self.biblioteca = Biblioteca()
        self.root = root
        self.root.title("Biblioteca")

        # Widgets para registrar libros
        tk.Label(root, text="Registrar Libro").grid(row=0, column=0, columnspan=2)
        tk.Label(root, text="Título").grid(row=1, column=0)
        self.titulo_entry = tk.Entry(root)
        self.titulo_entry.grid(row=1, column=1)
        tk.Label(root, text="ISBN").grid(row=2, column=0)
        self.isbn_entry = tk.Entry(root)
        self.isbn_entry.grid(row=2, column=1)
        tk.Label(root, text="Autor").grid(row=3, column=0)
        self.autor_entry = tk.Entry(root)
        self.autor_entry.grid(row=3, column=1)
        tk.Label(root, text="Categoría").grid(row=4, column=0)
        self.categoria_entry = tk.Entry(root)
        self.categoria_entry.grid(row=4, column=1)
        tk.Button(root, text="Registrar Libro", command=self.registrar_libro).grid(row=5, column=0, columnspan=2)

        # Widgets para registrar usuarios
        tk.Label(root, text="Registrar Usuario").grid(row=6, column=0, columnspan=2)
        tk.Label(root, text="Nombre").grid(row=7, column=0)
        self.nombre_usuario_entry = tk.Entry(root)
        self.nombre_usuario_entry.grid(row=7, column=1)
        tk.Label(root, text="Apellido").grid(row=8, column=0)
        self.apellido_usuario_entry = tk.Entry(root)
        self.apellido_usuario_entry.grid(row=8, column=1)
        tk.Label(root, text="ID Usuario").grid(row=9, column=0)
        self.id_usuario_entry = tk.Entry(root)
        self.id_usuario_entry.grid(row=9, column=1)
        tk.Button(root, text="Registrar Usuario", command=self.registrar_usuario).grid(row=10, column=0, columnspan=2)

        # Widgets para realizar préstamos
        tk.Label(root, text="Realizar Préstamo").grid(row=11, column=0, columnspan=2)
        tk.Label(root, text="Título Libro").grid(row=12, column=0)
        self.titulo_prestamo_entry = tk.Entry(root)
        self.titulo_prestamo_entry.grid(row=12, column=1)
        tk.Label(root, text="ID Usuario").grid(row=13, column=0)
        self.id_usuario_prestamo_entry = tk.Entry(root)
        self.id_usuario_prestamo_entry.grid(row=13, column=1)
        tk.Button(root, text="Realizar Préstamo", command=self.realizar_prestamo).grid(row=14, column=0, columnspan=2)

        # Widgets para devolver libros
        tk.Label(root, text="Devolver Libro").grid(row=15, column=0, columnspan=2)
        tk.Label(root, text="Título Libro").grid(row=16, column=0)
        self.titulo_devolucion_entry = tk.Entry(root)
        self.titulo_devolucion_entry.grid(row=16, column=1)
        tk.Label(root, text="ID Usuario").grid(row=17, column=0)
        self.id_usuario_devolucion_entry = tk.Entry(root)
        self.id_usuario_devolucion_entry.grid(row=17, column=1)
        tk.Button(root, text="Devolver Libro", command=self.devolver_libro).grid(row=18, column=0, columnspan=2)

        # Widget para mostrar libros
        tk.Button(root, text="Mostrar Libros", command=self.mostrar_libros).grid(row=19, column=0, columnspan=2)

    def registrar_libro(self):
        titulo = self.titulo_entry.get()
        isbn = self.isbn_entry.get()
        autor_nombre, autor_apellido = self.autor_entry.get().split()
        categoria_nombre = self.categoria_entry.get()

        autor = Autor(autor_nombre, autor_apellido)
        categoria = Categoria(categoria_nombre)
        libro = Libro(titulo, isbn, autor, categoria)

        self.biblioteca.registrar_libro(libro)

    def registrar_usuario(self):
        nombre = self.nombre_usuario_entry.get()
        apellido = self.apellido_usuario_entry.get()
        id_usuario = self.id_usuario_entry.get()

        usuario = Usuario(nombre, apellido, id_usuario)
        self.biblioteca.registrar_usuario(usuario)

    def realizar_prestamo(self):
        titulo = self.titulo_prestamo_entry.get()
        id_usuario = self.id_usuario_prestamo_entry.get()

        libro = next((l for l in self.biblioteca.libros if l.titulo == titulo), None)
        usuario = next((u for u in self.biblioteca.usuarios if u.id_usuario == id_usuario), None)

        if libro and usuario:
            self.biblioteca.realizar_prestamo(libro, usuario)
        else:
            messagebox.showerror("Error", "Libro o Usuario no encontrado.")

    def devolver_libro(self):
        titulo = self.titulo_devolucion_entry.get()
        id_usuario = self.id_usuario_devolucion_entry.get()

        prestamo = next((p for p in self.biblioteca.prestamos if p.libro.titulo == titulo and p.usuario.id_usuario == id_usuario), None)

        if prestamo:
            self.biblioteca.devolver_libro(prestamo)
        else:
            messagebox.showerror("Error", "Préstamo no encontrado.")

    def mostrar_libros(self):
        self.biblioteca.mostrar_libros()

if __name__ == "__main__":
    root = tk.Tk()
    app = BibliotecaApp(root)
    root.mainloop()
